### PR TITLE
[Sniper] Fix cache, and name matching tweak

### PR DESCRIPTION
**Pull request by Gobberwart**
## Short Description:

Two issues fixed here. 

1. Existing cache is not persistent, and gets reset every time the task runs, which results in repeated failed attempts to catch non-existent/otherwise failed pokemon. Even if cache hits, still generates an unnecessary log message. Changed to persistent cache, and check before attempting to snipe.

2. Name matching can be flaky depending on sources. Added method to manually fix names we know about, or use difflib (as with pokemonoptimizer) to try and match any we haven't already caught. Note that difflib alone is not sufficient for this, eg. it thinks source name "mr-mime" is a Grimer. Close but no cigar.

### CODE REVIEW

File: pokemongo_bot/cell_workers/sniper.py [hunk 1]
```diff
@@ -476,20 +499,25 @@ def _hash(self, pokemon):
...
+        if not self._is_cached(uniqueid):
             # Free space if full and store it
-            if len(self.cache) >= self.MAX_CACHE_LIST_SIZE:
-                self.cache.pop(0)
-            self.cache.append(pokemon)
+            if len(self.bot.sniper_cache) >= self.MAX_CACHE_LIST_SIZE:
+                self.bot.sniper_cache.pop(0)
+            self.bot.sniper_cache.append(uniqueid)
+
+    def _build_unique_id(self, pokemon):
```
Comment (YvesHenri) [inline pokemongo_bot/cell_workers/sniper.py]: This is totally unecessary. Use _hash instead.

File: pokemongo_bot/cell_workers/sniper.py [hunk 2]
```diff
@@ -152,10 +154,25 @@ def validate(self):
             raise
             
     def _fixname(self,name):
-        name = name.replace("mr-mime","mr. mime")
-        name = name.replace("farfetchd","farfetch'd")
+        if name:
+            name = name.replace("mr-mime","mr. mime")
+            name = name.replace("farfetchd","farfetch'd")
+            name = name.replace("Nidoran\u2642","nidoran m")
+            name = name.replace("Nidoran\u2640","nidoran f")
```
Comment (YvesHenri) [inline pokemongo_bot/cell_workers/sniper.py]: Seriously? Are you going to handle every name combination possible?

File: pokemongo_bot/cell_workers/sniper.py [hunk 3]
```diff
@@ -152,10 +154,25 @@ def validate(self):
...
         return name
 
+    def _get_closest_name(self, name):
```
Comment (YvesHenri) [inline pokemongo_bot/cell_workers/sniper.py]: What's this?

File: pokemongo_bot/cell_workers/sniper.py [hunk 4]
```diff
@@ -210,13 +227,12 @@ class Sniper(BaseTask):
     MIN_SECONDS_ALLOWED_FOR_CELL_CHECK = 10
     MIN_SECONDS_ALLOWED_FOR_REQUESTING_DATA = 5
     MIN_BALLS_FOR_CATCHING = 10
-    MAX_CACHE_LIST_SIZE = 200
+    MAX_CACHE_LIST_SIZE = 500
```
Comment (YvesHenri) [inline pokemongo_bot/cell_workers/sniper.py]: You  do realize people bot on raspberry devices, right? Increasing this this high might not be a good idea.

File: pokemongo_bot/cell_workers/sniper.py [hunk 5]
```diff
@@ -232,6 +248,9 @@ def initialize(self):
         self.altitude = uniform(self.bot.config.alt_min, self.bot.config.alt_max)
         self.sources = [SniperSource(data) for data in self.config.get('sources', [])]
 
+        if not hasattr(self.bot,"sniper_cache"):
+            self.bot.sniper_cache = []
```
Comment (YvesHenri) [inline pokemongo_bot/cell_workers/sniper.py]: What's this and what's it for?

File: pokemongo_bot/cell_workers/sniper.py [hunk 6]
```diff
@@ -311,60 +325,70 @@ def snipe(self, pokemon):
...
-            if exists:
-                self._log('Yay! There really is a wild {} nearby!'.format(pokemon.get('pokemon_name')))
-                self._teleport_back_and_catch(last_position, pokemon)
+            # Have we already tried this pokemon?
+            if not hasattr(self.bot,'sniper_unique_pokemon'):
+                self.bot.sniper_unique_pokemon = []
+            
+            # Check if already in list of pokemon we've tried
+            uniqueid = self._build_unique_id(pokemon)
+            if self._is_cached(uniqueid):
```
Comment (YvesHenri) [inline pokemongo_bot/cell_workers/sniper.py]: is_snipeable should check whether its been cached or not.

File: pokemongo_bot/cell_workers/sniper.py (hunk 2 above)
Comment (Gobberwart) [inline pokemongo_bot/cell_workers/sniper.py]: Read the description

File: pokemongo_bot/cell_workers/sniper.py (hunk 1 above)
Comment (Gobberwart) [inline pokemongo_bot/cell_workers/sniper.py]: How is it unnecessary? If there's an easy way to generate MD5 from string, then awesome. Please share your alternative code.

File: pokemongo_bot/cell_workers/sniper.py (hunk 3 above)
Comment (Gobberwart) [inline pokemongo_bot/cell_workers/sniper.py]: Again, read the description. Difflib is already used in pokemonoptimizer to TRY and get a real pokemon name from a supplied string. Eg. if source is "pigeot", it *should* return "pidgeot"... in many cases, it seems to work well.

File: pokemongo_bot/cell_workers/sniper.py (hunk 4 above)
Comment (Gobberwart) [inline pokemongo_bot/cell_workers/sniper.py]: You do realise that replacing dict with an md5 string makes it take FARRRRR less memory to store.

File: pokemongo_bot/cell_workers/sniper.py (hunk 5 above)
Comment (Gobberwart) [inline pokemongo_bot/cell_workers/sniper.py]: You had a "self.cache" which is nice but not persistent. Every time the task runs, it reset it to []. This creates a bot-global cache which will persist through task re-inits.

File: pokemongo_bot/cell_workers/sniper.py (hunk 6 above)
Comment (Gobberwart) [inline pokemongo_bot/cell_workers/sniper.py]: No need. It checks before it calls is_snipeable. Saves a wasted log message.

**Comment by Gobberwart** [comment]
OK @YvesHenri I've answered all your questions. Talk to me in chat if something is still unclear.

File: pokemongo_bot/cell_workers/sniper.py (hunk 6 above)
Comment (Gobberwart) [inline pokemongo_bot/cell_workers/sniper.py]: is_snipeable currently lumps everything together as a general fail. I've removed one of those for clarity.

File: pokemongo_bot/cell_workers/sniper.py (hunk 4 above)
Comment (Gobberwart) [inline pokemongo_bot/cell_workers/sniper.py]: But OK, sure... I'll compromise at 300.

File: pokemongo_bot/cell_workers/sniper.py (hunk 1 above)
Comment (Gobberwart) [inline pokemongo_bot/cell_workers/sniper.py]: Unless you're talking about the len(self.bot.sniper_cache) thing... which is your code so you tell me why it's necessary.

**Comment by Gobberwart** [comment]
Does anyone else have feedback for this?

**Comment by Gobberwart** [comment]
No feedback received and YH's requested changes are irrelevant. Merging.

**Pull request merged**